### PR TITLE
[WIP] Add PHP client

### PR DESCRIPTION
**Pull request by ackintosh**
With reference to java, work in progress now. 👀 

### TODO

- Fix build error
- Make PhpClientCodegen/AbstractPhpCodegen  DRY
- etc...

### CODE REVIEW

**Comment by jmini** [comment]
First idea: you should merge `master` to your feature branch or rebase all your commit on top of master.

This is what I did locally to review your code.

File: src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+import java.util.regex.Matcher;
+
+import io.swagger.v3.oas.models.media.*;
+import io.swagger.v3.parser.util.SchemaTypeUtil;
+import org.apache.commons.lang3.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractPhpCodegen extends DefaultCodegenConfig {
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java]: `DefaultCodegenConfig` is not imported. add following line:

    import io.swagger.codegen.languages.DefaultCodegenConfig;

File: src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+import java.util.regex.Matcher;
+
+import io.swagger.v3.oas.models.media.*;
+import io.swagger.v3.parser.util.SchemaTypeUtil;
+import org.apache.commons.lang3.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PhpClientCodegen extends DefaultCodegenConfig {
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java]: Same here: `DefaultCodegenConfig` is not imported.

File: src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java [hunk 3]
```diff
@@ -0,0 +1,731 @@
+package io.swagger.codegen.languages;
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java]: Package should be: `io.swagger.codegen.languages.php`

File: src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java [hunk 4]
```diff
@@ -0,0 +1,667 @@
+package io.swagger.codegen.languages;
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java]: Package should be: `io.swagger.codegen.languages.php`

File: src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+        } else if (!languageSpecificPrimitives.contains(type)) {
+            // type is a model class, e.g. User
+            example = "new " + getTypeDeclaration(type) + "()";
+        } else {
+            LOGGER.warn("Type " + type + " not handled properly in setParameterExampleValue");
+        }
+
+        if (example == null) {
+            example = "NULL";
+        } else if (Boolean.TRUE.equals(p.isListContainer)) {
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/AbstractPhpCodegen.java]: This produces a compile error by `p.isListContainer`.

Where did you get this file? [PhpClientCodegen](https://github.com/swagger-api/swagger-codegen/blob/3.0.0/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/PhpClientCodegen.java) on the `3.0.0` branch in the `swagger-codegen` has no copile error. You should have took this as base.

File: src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+        } else if (!languageSpecificPrimitives.contains(type)) {
+            // type is a model class, e.g. User
+            example = "new " + getTypeDeclaration(type) + "()";
+        } else {
+            LOGGER.warn("Type " + type + " not handled properly in setParameterExampleValue");
+        }
+
+        if (example == null) {
+            example = "NULL";
+        } else if (Boolean.TRUE.equals(p.isListContainer)) {
```
Comment (jmini) [inline src/main/java/io/swagger/codegen/languages/php/PhpClientCodegen.java]: This produces a compile error by `p.isListContainer`.

Where did you get this file? [PhpClientCodegen](https://github.com/swagger-api/swagger-codegen/blob/3.0.0/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/PhpClientCodegen.java) on the `3.0.0` branch in the `swagger-codegen` has no copile error. You should have took this as base.

**Comment by jmini** [comment]
I hope you will manage to remove the compile errors. Do not hesitate to ask if you need something.

**Comment by jmini** [comment]
In order to help you, I have started a wiki page:
[Swagger Codegen migration (swagger codegen generators repository)](https://github.com/swagger-api/swagger-codegen/wiki/Swagger-Codegen-migration-(swagger-codegen-generators-repository))

**Comment by HugoMario** [comment]
@ackintosh do you mind if we merge your fork with a different branch, in order to help with `TO-DO` tasks? or can you please grant access to `HugoMario` user in your forked repo?

**Comment by ackintosh** [comment]
Sorry for late reply 💦 
No, I don't. Please start your work in a different branch. ✨

**Comment by jmini** [comment]
I have automated a lot of the work that is requested to migrate the generators, this is no problem for me to migrate the PHP generator.

But I will not be able to evaluate the generated code and to confirm that it works as expected.

How do you want to proceed? Should I start something and you will test it? 
@HugoMario any other idea?

**Comment by HugoMario** [comment]
thanks @ackintosh 
@jmini sure, let me merge this into a new branch `php_client_generator`, so you can add your work there and i can later verify the generated code.

**Comment by jmini** [comment]
I have fetched the `php_client_generator` branch, but there is nothing more than on the `master` branch. I think you should merge `ackintosh/php-client` on your new branch.

What is the idea? 
Should I create a pull request for the new `php_client_generator` branch, should I push directly on the `php_client_generator` branch (but I think that I need a permission for that).

I can also fetch `ackintosh/php-client` locally, push the changes I had described in this merge request and then open a new one.

---

@HugoMario: I would appreciate if you could review and merge my existing pull requests for the `swagger-codegen-generators` and `swagger-codegen` repositories.

There is a lot important stuff going on in order to improve Swagger v3.

**Comment by HugoMario** [comment]
@jmini i already merge this PR against `php_client_generator` branch. So the idea is that any change you want to make, you can do it pointing to this branch.

**Comment by phpjob** [comment]
I'm currently checking a few options to generate an API client.
Is it currently possible for OAI 3.0? Or currently only for 2.0?

Thanks! :).

**Pull request merged**